### PR TITLE
Add SmallRye GraphQL extension

### LINKED ISSUE #2509: GraphQL
Hello, is it possible to use GraphQL and the GUI GraphiQL with Quarkus? @gsmet

### PR DESCRIPTION
**Pull request by phillip-kruger**
This PR adds Smallrye GraphQL (a MicroProfile GraphQL Implementation) to Quarkus.

This PR will also resolve #2509

### CODE REVIEW

**Comment by jmartisk** [comment]
As a side note, I have SmallRye Metrics integration with SmallRye GraphQL almost ready too, it should arrive soon after this is merged.

File: bom/runtime/pom.xml
Comment (gsmet) [inline bom/runtime/pom.xml:2998]: This one doesn't look necessary :).

File: extensions/pom.xml
Comment (gsmet) [inline extensions/pom.xml:166]: Indentation issues.

File: build-parent/pom.xml
Comment (gsmet) [inline build-parent/pom.xml:71]: Maybe add the `-api` suffix? Looks like they are all consistent except context-propagation.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (gsmet) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:72]: We don't do `@author` tags in Quarkus and rely on Git history.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (gsmet) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:81]: Better push that as a parameter of your methods. It's better to keep track of dependencies.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (gsmet) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:89]: I would make that a separate class.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (gsmet) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:151]: Is this related to SmallRye Metrics, should we test the presence of it?

**Comment by phillip-kruger** [comment]
Documentation PR here: https://github.com/quarkusio/quarkus/pull/8967

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (gsmet) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:186]: This is being worked on by @stuartwdouglas IIUC the issue.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (gsmet) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:212]: Why would the extension need SSL? It connects to something?

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (gsmet) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:261]: This should definitely be at the top of the class.

File: extensions/smallrye-graphql/runtime/pom.xml
Comment (gsmet) [inline extensions/smallrye-graphql/runtime/pom.xml:14]: This module needs a description, it's used in code.quarkus.io. Better start the description with a verb "Connect to..." or whatever :).

File: extensions/smallrye-graphql/deployment/pom.xml
Comment (gsmet) [inline extensions/smallrye-graphql/deployment/pom.xml:41]: The jsonp deployment is missing here. It should be symmetrical with the runtime pom. As this dependency comes with jsonb, I would simply remove it from the runtime module

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java
Comment (gsmet) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:106]: Please put the constants at the top of the class.

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java
Comment (gsmet) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java:18]: Let's drop useless javadoc comments.

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java
Comment (gsmet) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java:26]: This is incorrect. You probably want a `BeanContainerListener`.

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/QuarkusClassloadingService.java
Comment (gsmet) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/QuarkusClassloadingService.java:17]: This needs @stuartwdouglas 's attention.

File: tcks/microprofile-graphql/pom.xml
Comment (gsmet) [inline tcks/microprofile-graphql/pom.xml:44]: Please make the indentation consistent.

**Review by gsmet — CHANGES_REQUESTED** [review]
Cool! I added a bunch of comments, mostly on the form. Could you address them so that we get these out of the way and can focus on the implementation?

Thanks!

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (jmartisk) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:151]: Metrics integration is not supposed to be covered by this PR yet, I am planning to add that later, so I think this is some leftover that can be removed for now

Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:151]: No, I added it to load the SPI (Native mode). @jmartisk will this use another service implementation ? I can remove this, then you add it when you do the metrics ?

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:186]: Yes, once his changes is in I can remove this and the special QuarkusClassloadingService class

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (jmartisk) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:151]: Well, yeah, this is just one part of the metrics integration - without the rest, the metrics won't work properly anyway. So yeah I'd say we can remove these six lines for now, I'll do metrics separately later and re-add this

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/QuarkusClassloadingService.java
Comment (phillip-kruger) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/QuarkusClassloadingService.java:17]: Yes, this is the hack until @stuartwdouglas 's fix is in. Same as open-api extension. I'll update both once the fix is in.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (jmartisk) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:315]: Could we use `org.apache.commons.io.IOUtils.toByteArray(InputStream)` instead? It seems to be somewhere among the dependencies and some other extensions use it too

Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:307]: Happy to use that, did not want to introduce an new dependency, but if it's there it makes sense. Let me know.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (jmartisk) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:83]: I am getting a compile error now, seems you forgot to add this class after extracting it

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (gsmet) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:307]: Please check before doing that that the dependency is not there by pure luck (i.e. brought by another one that is not really related to what we do and where it could go away)

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:83]: Sorry. (But it build on my machine :))
I'll commit now with the other Arc fix

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java
Comment (phillip-kruger) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java]: Thanks :)

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:85]: @jmartisk please check now

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:317]: Seems like it's available:
GroupId: commons-io 
ArtifactId: commons-io 
Version: 2.6 
Type: jar Scope: compile 

Via: maven-shared-utils 
Via: maven-embedder 
Via: quarkus-bootstrap-core 
Via: quarkus-core-deployment

Should I go ahead and use that ? @gsmet ?

**Comment by jmartisk** [comment]
No further issues from my side at the moment. Tested it with the latest GraphQL snapshot and it works fine, except the introspection in UI in native mode, but I don't really know how to fix that.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (gsmet) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:317]: Given it's coming incidentally from Maven, I wouldn't use it.

Better add what you need to `io.quarkus.deployment.util.IoUtil`.

Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:317]: No need to add anything, the method was there already. Doing a fix now

**Comment by phillip-kruger** [comment]
Ready for review :)

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/QuarkusClassloadingService.java
Comment (phillip-kruger) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/QuarkusClassloadingService.java:15]: Can I resolve this ? As I said as soon as there is a fix I will remove the hack here and in open-api ? @gsmet  ?

**Comment by gsmet** [comment]
@phillip-kruger is there anything still not working? Not saying it should block this PR. I just would like to have a clear status?

Could you squash all the commits into one?

**Comment by phillip-kruger** [comment]
Yes, there is a weird intermittent error after a hot reload. I am looking at it now, with the help of clement.
Then the UI is not working properly in native mode. You can do queries, but introspection is not working. I will look at that too a.s.a.p

**Comment by phillip-kruger** [comment]
@gsmet :  Squashed ! :)

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (stuartwdouglas) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java:341]: This should be in a try with resources block.

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java
Comment (stuartwdouglas) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:65]: I don't think you need to bother with the status messages, they are not supported with HTTP/2 and are generally considered to have been a bad idea. We don't want to encourage their use as they are basically deprecated.

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java
Comment (stuartwdouglas) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:89]: You probably want to cache this in a volatile variable. CDI.current().select() is relatively expensive.

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java
Comment (stuartwdouglas) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java:21]: Why do you pass the bean class in like this, when it is hard coded anyway? Why not just use GraphQLProducer.class?

**Review by stuartwdouglas — DISMISSED** [review]

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java
Comment (mkouba) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java:89]: In fact, it does not have to be a `volatile` field but a final field initialized in the constructor. Also `Arc.container().instance()` is a bit more effective.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java
Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallryeGraphqlProcessor.java]: Cool, thanks. Did not know about this.

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java
Comment (phillip-kruger) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLExecutionHandler.java]: Did a lazy load with volatile variable.

**Comment by gsmet** [comment]
I don't see any integration tests or am I missing something? We need some to get native covered by our CI. Also it needs to be added to the github CI workflow.

I don't think the TCK tests cover native, right?

**Comment by phillip-kruger** [comment]
@gsmet - I fixed all the review items from @stuartwdouglas and @mkouba , then squashed, but made a mistake (fixed now) - however @boring-cyborg added a bunch of labels after my mistake, that I can not remove. 

Can you please remove the following labels:  

- area/arc 
- area/devtools 
- area/documentation 
- area/maven 
- area/mongodb 
- area/panache 
- area/gradle  

Thanks and apologies.

**Comment by gsmet** [comment]
OK, I think the only thing blocking this PR is that we need to have test coverage for native. I haven't seen an answer to my question? Right now, I don't think it's covered.

**Comment by phillip-kruger** [comment]
Hi @gsmet - Busy adding this as we speak. Should have it ready later today

**Comment by phillip-kruger** [comment]
@gsmet - Integration tests with native added. Please let me know if this is ok. Thanks :)

File: integration-tests/smallrye-graphql/nbactions.xml
Comment (jmartisk) [inline integration-tests/smallrye-graphql/nbactions.xml:1]: This file shouldn't be here

**Review by jmartisk — CHANGES_REQUESTED** [review]

File: integration-tests/smallrye-graphql/nbactions.xml
Comment (phillip-kruger) [reply integration-tests/smallrye-graphql/nbactions.xml]: Sorry about this - removed the file and added Netbeans files to gitignore

**Review by jmartisk — APPROVED** [review]

**Comment by gsmet** [comment]
@phillip-kruger I added a commit to add the native tests to the CI workflow. It's not something automatic. Let's see what CI has to say and then merge.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
I fixed a conflict in the CI set up and force pushed. Hopefully I will be able to merge it right after the CI run.

**Comment by gsmet** [comment]
@phillip-kruger @jmartisk merged, thanks! I let you continue with follow-up PRs?

**Comment by gsmet** [comment]
Also, we need a proper guide and quickstart. I know there is some work in progress but we need to finish that work before the end of the month.

**Comment by phillip-kruger** [comment]
Thanks @gsmet . Yes we are working on guide and quick start with @cemnura . Will get it done by the end of the month. Thanks!

**Pull request merged**
